code/modules/clothing/glasses/engine_goggles.dm

```diff
@@ -0,0 +1,76 @@
+//Engineering Mesons
+
+/obj/item/clothing/glasses/meson/engine
+	name = "Emgineering Meson Scanner"
+	desc = "Goggles used by engineers. The Meson Scanner mode lets you see basic structural and terrain layouts through walls, regardless of lighting condition. The T-ray Scanner mode lets you see underfloor objects such as cables and pipes."
+	icon_state = "meson"
+	item_state = "glasses"
+	origin_tech = "magnets=2;engineering=2"
+	darkness_view = 1
+	vision_flags = SEE_TURFS
+	invis_view = SEE_INVISIBLE_MINIMUM
+	var/state = 0	//0 - regular mesons mode	1 - t-ray mode
+	var/invis_objects = list()
+	action_button_name = "Change scanning mode"
+
+/obj/item/clothing/glasses/meson/engine/attack_self(mob/user)
+	ui_action_click()
+
+/obj/item/clothing/glasses/meson/engine/ui_action_click(mob/user)
+	state = !state
+	if(state)
+		SSobj.processing.Add(src)
+		vision_flags = 0
+		darkness_view = 2
+		invis_view = SEE_INVISIBLE_LIVING
+	else
+		vision_flags = SEE_TURFS
+		darkness_view = 1
+		invis_view = SEE_INVISIBLE_MINIMUM
+		invis_update()
+	update_icon()
+
+/obj/item/clothing/glasses/meson/engine/process()
+	if(!state)
+		SSobj.processing.Remove(src)
+		return null
+	else
+		var/mob/living/carbon/human/user = loc
```

Comment:
the glasses don't stop processing if you drop them afaik, so every time you use user = loc, it's gonna runtime when calling user.glasses.